ClinGen/clincoded
# Connects to #683. VCI population tab UI.

This PR consists of changes pertinent to pulling in external data and populate the tables with the data on the Population tab.

**Known issues:**
1. Currently the default active tab is "Basic Information". When the page loads, the app fetches data needed for the "Basic Information" tab. However, if I subsequently click on the "Population" tab, the expected data does not appear to have been pre-fetched (or fetched at the moment when the "Population" tab becomes active). I suspect that Min's change in which clicking a tab passes the tab index to the URL as a param would potentially resolve this issue. For the time being, the hacky way to fetch data for the "Population" tab is to hard-code the "selectedIndex" (e.g. selectedIndex={1}) in the "VariantCurationInterpretation" component (e.g. interpretation.js).
2. Previously I stored all the external data in React states. With the react-localstorage npm, the states were kept in the browser's localStorage and persistent. Simply refreshing the page would not invoke fetching. With my latest commits, in which I created a global population object to store the ExAC and ESP data (for the time being) instead of using states, the data is no longer being kept in the browser's localStorage via the npm package. As such, refreshing the page would invoke fetching and it results in a noticeable delay in displaying data in the ExAC and ESP tables.
3. Likely due to the way that the react-localstorage npm package has been implemented, changing the ClinVar variation id fails to force the data to be refreshed for the display. For now, we can manually delete the browser's localStorage to see new data in the tables.

**Technical notes:**
1. Added a JSON mapping file (e.g. NC_genomic_chr_format.json) consisting of NC_genomic and chr pairs. This mapping is provided by Matt and used for constructing variant id in retrieving data from myvariant.info.
2. Added more comments in the parse-resources.js and basic_info.js files. Also replaced most of the one-letter variables with more descriptive names in these files.
3. In the population.js, a global object has been created to store all the externally pulled in data instead of keeping the data in React states. The intent is to have a centralized source for the displaying data in the tables, saving the population data object into the database, as well as acting as the original source for comparing any potential changes in the newly pulled in data. The properties of this global object currently have `null` values, but further development shall enable these properties to contain pre-existing values from the database if there are any.
4. Due to time constraint, only the ExAC and ESP tables have been refactored to get their data from the global object. And with the use of the global object, some states previously associated with these 2 tables may no longer be needed.

**Step to test this PR:**
1. Remember to set the `selectedIndex` to 1 in the "VariantCurationInterpretation" component and rerun `bin/grunt`.
2. Login and go to http://localhost:6543/select-variant/.
3. Add a ClinVar variation ID - 139214.
4. Upon arriving at the Population tab, expect to see data being displayed in the 3 tables.
5. Other ClinVar variation IDs can also be used for testing purpose: 3520 or 55962 (remember to clear the browser's localStorage before testing a new variation ID).
